### PR TITLE
added visitor & front-end code for PR, no generated code

### PR DESCRIPTION
**Pull request by Groostav**
This is a PR to demo the new babel code as ported out of our codebase. Note that it is largely read-only, and that any issues from this PR should be written as new issues.

Some notes about the first implementation:
- it uses ANTLR to generate a parser. At this time, master is at cae22712324134e60d3118f9829b7fc2f7b0e8b9, where that generated code is simply checked-in to the codebase. I hope to remove this soon.
- We use a couple of "walkers" aka "listeners" aka "visitors" to do a post-order traversal of the tree to validate it, determine its use (boolean or scalar), and generate a runtime for it.
- The generated runtime simply hits the java runtime for all of its operations, using a `LinkedList` backed `Deque`, and an _ordered_ `Map` for global variables.

This implementation is fully immutable and thus fully concurrency-safe --unlike its predecessor.

Some future work:
- I'm looking to expose the AST for further analysis by tools like [sojourn-CVG](../../../groostav/sojourn-CVG), 
- I'm not happy with some of the names of data classes, so those are likely to change.

### CODE REVIEW

File: src/test/kotlin/com/empowerops/babel/DemoKT16752.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.empowerops.babel
+
+import org.testng.annotations.Test
+
+typealias MyUnaryOp = (Double) -> Double
+object DelegatedOp: MyUnaryOp by Math::cos
+
+class DemoKT16752 {
+
+    @Test(enabled = false)//demonstrates bug KT16752 => expects failure while that bug exists
```
Comment (NavalOrange) [inline src/test/kotlin/com/empowerops/babel/DemoKT16752.kt:10]: izzis a kotlin bug?

**Review by NavalOrange — COMMENTED** [review]
Hey Geoff, This is a lot of stuff and most of it just looks like its copied over from OASIS, could you outline what you changed to get the speedups?

**Comment by Groostav** [comment]
The single biggest change was in code generation:

where we used to have a `SymbolTable` that indexed values by the node on the tree that declared them, we now have a _much_ more elegant `RuntimeMemory` object. It has three fields, I'll note them on the object itself.

## The `CodeGeneratingWalker` ##

The other change is in the `CodeGeneratingWalker` itself. It use to be java _with no if-statements_, it is now more regular & idiomatic kotlin, albeit with something of a DSL.

>_how the hell did you write it with no if-statements_

When I wrote the original code generator years ago I was big on the theory that _bugs come from if statements, and I can express my if-statements through polymorphism_ so, like any idiot, I made everything polymorphic. Believe it or not I more-or-less achieved this and for a while I was proud of my accomplishment.

Then I had to make some modifications...



Consider:

```kt
val operations: Stack<(Double, Double) -> Double> = emptyStack()
val values: Stack<Double> = emptyStack()

fun exitExpression(exprContext: ExprContext){
  val operation = operations.pop()
  val right = values.pop()
  val left = values.pop()

  val newValue = operation.invoke(left, right)
  values.push(newValue)
}
fun exitPlus(plusContext: PlusContext){
  operations += { left, right -> left + right }
}
fun exitMinus(plusContext: MinusContext){
  operations += { left, right -> left - right }
}
fun exitValue(valueContext: ValueContext){
  values += valueContext.text.toDouble()
}
```

vs

```kt
val values: Stack<Double> = emptyStack()

fun exitExpression(exprContext: ExprContext){
  val right = values.pop()
  val left = values.pop()

  val newValue = when(exprtContext.operatorChild()){
    is PlusContext -> left + right
    is MinusContext -> left - right
  }

  values.push(newValue)
}
fun exitValue(valueContext: ValueContext){
  values += valueContext.text.toDouble()
}
```

notice that the top doesn't have any if statements where the bottom does (if we consider `when` statements to be `if` statements, which they are, just with some syntactic sugar on them). But I think the bottom is a whole lot clearer, and more difficult to get wrong --it might be possible for some jackass to copy-and-paste in an implementation for multiplication and division, or some other boolean operator, and forget to update his expression with `*` and `/` instead of `+` and `-`, but this is true of the no-~~if~~when-statement top implementation too. 

Interestingly enough, the top is also slower, since it requires more complex object allocations (building a closure, including loading an anonymous class, calling its constructor, etc, while pretty fast, is still one of the slowest operations here) and many more function calls. The later, being a switch, can be heavily optimized.

The actual implementation changes I made are different --and they still rely on an `operations` and `values` stack--, but they're along these lines.

## The SymbolTable ##

The other big change I made was to the SymbolTable. We used to tie values to nodes on the parse tree, meaning we had to keep the parse tree around at runtime (which is more or less irrelevant since the [antlr parse tree nodes](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/ParserRuleContext.java) is very very compact), and we also _had to traverse the tree at runtime to find values_. This tree traversal, while always running in log-n complexity of the expression (ie fast), was bug prone and required a number of slow operations with the hashmap.

This was replaced with a vastly simpler strategy using a `Map<String, Double>` for global variables (the variables passed in at `evaluate` time), a `Stack<Map<String, Double>>` for a heap which is tolerant of nested scopes (eg `"x1 + sum(1, 2, x1 -> x1) + x1".evalute("x1" to 100)` returns `203`, note that `x1` must be re-assined a value but in a nested scope), and a simple `Stack<Double>` for local values executed by simple expressions. 

This means that the vast majority of value lookups are either a `stack.pop` call or a `globalVars[name]` call. We do sometimes have to traverse up through the nested scopes, but the time that call takes no longer involves complex searching logic, instead its much simpler and proportional to the number of nested scopes, rather than the height of the expression tree. Simpler & faster.

File: src/test/kotlin/com/empowerops/babel/DemoKT16752.kt (hunk 1 above)
Comment (Groostav) [reply src/test/kotlin/com/empowerops/babel/DemoKT16752.kt:10]: yes

**Comment by Groostav** [comment]
this got merged in a while ago, certainly as of 0.11, I just forgot about this PR.

**Pull request closed without merging**